### PR TITLE
Switch to golangcli lint

### PR DESCRIPTION
**Pull request by manoranjith**
This PR addresses issue #4.

### CODE REVIEW

**Comment by danielksan81** [comment]
Running make lint leads to the following error output:
`
----------------------------------------------------------
                  Installing Dependencies                 
----------------------------------------------------------

govendor...	Successful
golangci-lint v1.21.0...	Successful

----------------------------------------------------------
                     Linting the code                     
----------------------------------------------------------

Running Linters...	Failed. Below warnings and errors are reported

level=error msg="Running error: context loading failed: package github.com/stretchr/testify/require: no go files to analyze"

Makefile:30: recipe for target 'lint' failed
make: *** [lint] Error 1
`

**Comment by danielksan81** [comment]
> 
> 
> Running `make lint` leads to the following error output:
> 
> Running Linters... Failed. Below warnings and errors are reported
> 
> level=error msg="Running error: context loading failed: package github.com/stretchr/testify/require: no go files to analyze"
> 
> Makefile:30: recipe for target 'lint' failed
> make: *** [lint] Error 1

At least I had to delete the folder build/workspace_/ manually to get it run without any error.

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,287 @@
+# options for analysis running
+run:
+  # include test files or not, default is true
+  tests: true
+
+  skip-dirs:
+    - src/external_libs
+    - autogenerated_by_my_lib
```
[the remark this answers is not in the document]
Comment (choeppler) [reply .golangci.yml]: Curiously enough that directory is to be excluded on the commandline, cf. https://github.com/direct-state-transfer/dst-go/pull/14/files#diff-75e934896def206c5aeb231d2f5c9cb9R313

**Comment by choeppler** [comment]
Trying to reproduce the error @danielksan81 reported [above](https://github.com/direct-state-transfer/dst-go/pull/14#issuecomment-551750848), I started with a clean copy of the dst-go repo:
```bash
$ git clone https://github.com/boschresearch/dst-go.git
$ git checkout switch-to-golangcli-lint 
$ make lint
go build -o build/build ./build
build/setEnv.sh build/build lint  
Creating a fake go workspace, so that user can build from source placed in non-go directory
Temp GOPATH set to fake go workspace : /home/hpc2si/share/work/perun/tmp/tmp/dst-go/build/workspace_/src/github.com/direct-state-transfer/dst-go 


----------------------------------------------------------
                  Installing Dependencies                 
----------------------------------------------------------

govendor...	Successful
golangci-lint v1.21.0...	Successful

----------------------------------------------------------
                     Linting the code                     
----------------------------------------------------------

Running Linters...	Failed. Below warnings and errors are reported

level=error msg="Running error: context loading failed: package github.com/ethereum/go-ethereum/common: no go files to analyze"

Makefile:30: recipe for target 'lint' failed
make: *** [lint] Error 1
```
The above make error can be fixed by first running `make install`. Then `make lint` runs successfully.

The error reported by @danielksan81 mentioned `github.com/stretchr/testify/require`. However, after running
```
make install
make test BUILDOPTS="-v -short"
make runWalkthrough BUILDOPTS="--simulated_backend"
```
`require` isn't present in my workspace (obsolete dependency?) and `make lint` still runs without any errors.

So, it works for me... Can you check whether you can still reproduce the behavior you've been observing in your setup, @danielksan81 ?

**Comment by manoranjith** [comment]
Thanks for testing it out. I have figured out how to exactly reproduce the error and the fix. 

The error is caused, when the dependencies are missing (as it is the case immediately after clean checkout) and `make lint` command is invoked. 

Alternatively when `make lint` is invoked after `make install` or `make test` or `make runWalkthrough` as @choeppler has mentioned,  it completes successfully because the dependencies are fetched by any of the previous command.

Fix is to add [fetchVendoredPackages()](https://github.com/direct-state-transfer/dst-go/blob/develop/build/ci.go#L143) before invoking the linter in [build/ci.go line #91](https://github.com/direct-state-transfer/dst-go/blob/develop/build/ci.go#L91) like it is done for other make commands.

Interestingly, the previous metalinter - "gometalinter.v2" did not require the dependencies to be present, but this one "golangci-iint" requires it to be. 

So, I am still trying to understand, which of the linters invoked by golangci-lint is requiring the dependencies to be present.

**Comment by danielksan81** [comment]
> So, it works for me... Can you check whether you can still reproduce the behavior you've been observing in your setup, @danielksan81 ?

I have the same behavior, it is reproducable in a go workspace as well outside. After running `make install` the linting works fine.
A general doubt: Does it make sense in the build scripts to clean the build/workspace folder before performing the task to execute? Then it is always a new build without any side effects from previous builds.

**Comment by choeppler** [comment]
> A general doubt: Does it make sense in the build scripts to clean the build/workspace folder before performing the task to execute? Then it is always a new build without any side effects from previous builds.

Well, I've been pondering this question as well. What really bugs me with the current build script
- no offline build possible; it always tries to download govendor and golangci-lint
- the standard build targets (install, test, ...) additionally always try to fetch all dependencies.
- there are the ciBuild targets which do not re-fetch the dependencies but still try to download and install the above mentioned tools. So even those targets don't support a quick modify-test cycle.

While there is some merit in having those "clean setup build"-targets, access to more granular targets would be quite useful. And would it suffice to fetch dependencies, when `vendor.json` changes, for example?

File: .golangci.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (manoranjith) [reply .golangci.yml]: Updated doc.go in this [commit](https://github.com/direct-state-transfer/dst-go/pull/14/commits/ef8197bc0e7c88ef38176dae2a8ec7460bbc9e18)

**Comment by manoranjith** [comment]
Updated the lint target in make to fetch vendored depedencies before invoking the make command in this [commit](https://github.com/direct-state-transfer/dst-go/pull/14/commits/2b0e75743feecdd41403444d09a63ec3d7d02c8a).

File: build/doc.go [hunk 2]
```diff
@@ -22,7 +22,8 @@ a go development environment or using existing one for this project is not desir
 Available commands are:
 
 	install        -- install the dst-go software
-	lint           -- run meta-linter with configuration linterConfig.json
+	lint           -- run meta-linter with configuration as defined in the
+					  .golangci.yml file in project root.
```
Comment (choeppler) [inline build/doc.go]: I'd prefer a short one-line message as in the original version, i.e. something like
```
lint           -- run meta-linter with configuration 'dst-go/.golangci.yml'
```

**Comment by choeppler** [comment]
It would be easier to follow the changes if you don't force update during the review but wait until we're ready to merge. Then squashing and force updating before merging is fine IMO...

File: build/doc.go (hunk 2 above)
Comment (manoranjith) [reply build/doc.go]: Changes updated in this [commit](https://github.com/direct-state-transfer/dst-go/pull/14/commits/49c73432690b6042f1b8fd998b95db4f8a0d4a5e). I have also updated the message for runWalkthrough target to fit in one line.

File: build/doc.go [hunk 3]
```diff
@@ -22,11 +22,9 @@ a go development environment or using existing one for this project is not desir
 Available commands are:
 
 	install        -- install the dst-go software
-	lint           -- run meta-linter with configuration linterConfig.json
+	lint           -- run meta-linter with configuration "dst-go/.golangci.yml"
 	test           -- run unit tests, with options as in "go test" command
-	runWalkthrough -- build and run walkthrough. Behavior is configurable with
-	                  options such as "--simulated_backend", "--real_backend",
-	                  etc.
+	runWalkthrough -- build and run walkthrough, use "-h" flag to see options
```
Comment (choeppler) [inline build/doc.go]: So, this message appears when one invokes `make help`. When I call `make -h` or `make runWalkthrough -h` as advised here, the output is
```
make runWalkthrough -h
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
<snip />
```
Do you mean invoke with `BUILDOPTS="-h"` to see options?

Comment (manoranjith) [reply build/doc.go]: Yes, as it is the case for for the other options mentioned in `make help`. 
This information can be added in make help message in a follow up issue.

**Comment by choeppler** [comment]
Hi @manoranjith, thanks for the contribution. As a final step before merging the changes into develop, please squash the commits and force push, so you can provide a consolidated commit message and we'll end up with clean commit history.

**Pull request merged**